File: arrays-and-hashing/k-frequent-elements.py
def k_frequent_elements(nums, k):
     """
     Returns the k most frequent elements in the given list.

     Args:
          nums (List[int]): The input list of integers.
          k (int): The number of most frequent elements to return.

     Returns:
          List[int]: The k most frequent elements in the input list.
     """
     hashMap = {}
     arr = [[] for i in range(len(nums) + 1)]

     for i in nums:
          hashMap[i] = 1 + hashMap.get(i, 0)

     for n, c in hashMap.items():
          arr[c].append(n)
          
     result = []
     for i in range(len(arr) - 1, 0, -1):
          for n in arr[i]:
               if (len(result) != k):
                    result.append(n)
               
     return result

nums = [1,2,2,3,3,3]
k = 2
print(k_frequent_elements(nums,k))

File: arrays-and-hashing/java-solutions/KFrequentElements.java
import java.util.*;

public class KFrequentElements {
     public static void main(String[] args) {
          System.out.println(kFrequentElements(new int[] { 1, 2, 2, 3, 3, 3 }, 2));
          return;
     }

     public static List<Integer> kFrequentElements(int[] elements, int k) {
          // stores numbers and their frequencies
          HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();

          // used to index a number by it's frequency
          List<List<Integer>> arr = new ArrayList<List<Integer>>();
          // initialize the array
          for (int i = 0; i < elements.length; i++) {
               arr.add(new ArrayList<>(Arrays.asList()));
          }

          // populate the map
          for (int element : elements) {
               if (map.containsKey(element)) {
                    map.put(element, map.get(element) + 1);
               } else {
                    map.put(element, 1);
               }
          }

          // populate the array from the map
          for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
               int key = entry.getKey();
               int frequencyIndx = entry.getValue();

               arr.get(frequencyIndx).add(key);
          }

          // final result
          List<Integer> result = new ArrayList<Integer>();

          // find the 'k' most frequent elements in the array
          for (int i = arr.size() - 1; i > -1; i--) {
               List<Integer> subArr = arr.get(i);
               for (int j = 0; j < subArr.size(); j++) {
                    if (result.size() != k) {
                         result.add(subArr.get(j));
                    }
               }
          }
          return result;
     }
}
